### ISSUE TITLE
iOS issue reported for "App Hanging" but it's just the app going to the background

### ISSUE THREAD

**Issue by acoutts** [getsentry/sentry-dart#1284]
_Platform:_
- [ ]  Dart
- [X]  Flutter Android or iOS
- [ ]  Flutter Web

_IDE:_
- [X]  VSCode
- [ ]  IntelliJ/AS
- [ ]  XCode
- [ ]  Other, which one?

_split-debug-info and obfuscate (Flutter Android or iOS) or CanvasKit (Flutter Web):_
- [ ]  Enabled
- [X]  Disabled

_Platform installed with:_
- [X] pub.dev
- [ ] GitHub

Output of the command `flutter doctor -v` below:

```
[✓] Flutter (Channel beta, 3.7.0-1.5.pre, on macOS 13.2 22D49 darwin-arm64, locale en-US)
    • Flutter version 3.7.0-1.5.pre on channel beta at /Users/andrew.coutts/Projects/flutter
    • Upstream repository git@github.com:flutter/flutter.git
    • Framework revision 099b3f4bf1 (3 weeks ago), 2023-01-20 18:35:12 -0800
    • Engine revision 45c5586f2a
    • Dart version 2.19.0 (build 2.19.0-444.6.beta)
    • DevTools version 2.20.1

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0-rc2)
    • Android SDK at /Users/andrew.coutts/Library/Android/sdk
    • Platform android-33, build-tools 33.0.0-rc2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.2)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14C18
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2021.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.13+0-b1751.21-8125866)

[✓] VS Code (version 1.75.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.58.0

[✓] VS Code (version 1.76.0-insider)
    • VS Code at /Applications/Visual Studio Code - Insiders.app/Contents
    • Flutter extension version 3.58.0

[✓] Connected device (4 available)
    • V2135 (mobile)             • 13755444730009Q                      • android-arm64  • Android 11 (API 30)
    • iPhone 14 Pro Max (mobile) • 3339E92F-270A-478D-B9E1-B5C143E3B6A4 • ios            • com.apple.CoreSimulator.SimRuntime.iOS-16-2 (simulator)
    • macOS (desktop)            • macos                                • darwin-arm64   • macOS 13.2 22D49 darwin-arm64
    • Chrome (web)               • chrome                               • web-javascript • Google Chrome 109.0.5414.119

[✓] HTTP Host Availability
    • All required HTTP hosts are available

• No issues found!
```

The version of the SDK (See pubspec.lock):
**4.x.x**

---
I have the following issue:

I'm getting a lot of reports like this saying the app is hanging but I don' think it is, it's just the app going to the background.

<img width="1090" alt="Screenshot 2023-02-14 at 7 45 41 AM" src="https://user-images.githubusercontent.com/9597207/218742646-794614e2-522b-4a90-b5d9-9401b919b6b9.png">

**Steps to reproduce:**
Unknown, came in via sentry.

**Actual result:**
Seeing false issues.

**Expected result:**
Should only report hangs when it actually hangs

**Comment by krystofwoldrich** [comment]
Thanks for the message we will look into it.

Have you tried to reproduce it in a simulator/on your own device?

**Comment by acoutts** [comment]
Yeah I can't seem to find an exact replication - i think because it can be random when the OS actually disposes of your app in the background.

**Comment by marandaneto** [comment]
cc @philipphofmann or @brustolin

**Comment by philipphofmann** [comment]
The AppHang tracker ignores AppHangs if the app is in the background, see https://github.com/getsentry/sentry-cocoa/blob/a91f733e4fffe80a74166aa035801fb7039a6762/Sources/Sentry/SentryANRTracker.m#L115-L118

This logic depends on `UIApplicationDidEnterBackgroundNotification` and `UIApplicationWillEnterForegroundNotification`, see https://github.com/getsentry/sentry-cocoa/blob/a91f733e4fffe80a74166aa035801fb7039a6762/Sources/SentryCrash/Recording/SentryCrash.m#L560-L568

Could it be that these notifications are not posted adequately on a Flutter app, @marandaneto? 

Maybe your app is hanging while it's going to the background, @acoutts. Could that be the case?

**Comment by marandaneto** [comment]
> Could it be that these notifications are not posted adequately on a Flutter app, @marandaneto?

I don't think Flutter interferes with that, the notifications are an OS feature.

**Comment by marandaneto** [comment]
@acoutts any update on that?

> Maybe your app is hanging while it's going to the background, @acoutts. Could that be the case?

**Comment by acoutts** [comment]
I just dont see how it could hang, the only thing I have added is this:
<img width="841" alt="Screenshot 2023-03-08 at 9 17 37 AM" src="https://user-images.githubusercontent.com/9597207/223736444-36a88a1a-b0f5-48a2-a759-e2445864a558.png">

This lets us overlay an image over the app for privacy when the app is backgrounded.

**Comment by acoutts** [comment]
Otherwise it's just a normal flutter app so any hanging behavior must be coming from the flutter framework (or detected wrong).

**Comment by marandaneto** [comment]
Could be that after that (presenting the new screen), the app gets another `UIApplicationWillEnterForegroundNotification`? or if by presenting a screen during `UIApplicationDidEnterBackgroundNotification` actually hangs the main thread? app is in the background but you actually want to draw a screen?

**Comment by marandaneto** [comment]
We'll try to reproduce this, we'll get back to you, thanks.

**Comment by brustolin** [comment]
Hello @acoutts, why do you think the app is moving to the background? Because of the breadcrumbs?

The app hanging detection is really good, if it reports an app hang is because the main thread is hanging. But the stack trace that is captured may be drifting. 

Maybe the app was hanging, the user switch apps and when we captured the stacktrace the app was already in the background.

I cant reproduce, and the image don't give me much information.

**Comment by acoutts** [comment]
Thanks for confirming. There must be something wrong with how I've setup my app then.

I have 2 app hang events - one is in the original screenshot, here is another one we are getting a lot of (300+ users):
![Screenshot 2023-03-14 at 3 26 17 PM](https://user-images.githubusercontent.com/9597207/225115559-c105d529-04fa-4e10-812d-b0a5d9fa248b.png)

**Comment by brustolin** [comment]
Thanks for the quick reply. 

The problem, is that the main thread may have a lot of work scheduled, we detect an app hang, and it happens that the stacktrace was in that particular point. 

The source of the problem may be elsewhere. Hard to tell. But we will keep investigating to try to improve this.
